.github/workflows/cypress.yml

```diff
@@ -0,0 +1,87 @@
+name: Cypress Tests
+
+on:
+  pull_request:
+    types: [opened]
+  push:
+    branches:
+      - '**'
+
+jobs:
+  # TODO: remove once speed/batch size testing complete
+  # interval of 6 files:
+  #       specs: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21 ]
+  # interval of 5 files:
+  #       specs: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25 ]
+  # interval of 4 files:
+  #       specs: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31 ]
+  # interval of 3 files:
+  #       specs: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41 ]
+
+  e2e_tests:
+    
+    runs-on: ubuntu-latest
+
+    permissions:
+      actions: read
+
+    defaults:
+      run:
+        shell: bash -ileo pipefail {0}
+
+    strategy:
+      fail-fast: false
+      matrix:
+        specs: [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21 ]
+    
+    steps:
+      - name: Checkout Streamlit code
+        uses: actions/checkout@v3
+        with: 
+          persist-credentials: false
+          submodules: 'recursive'
+      - name: Set up Python 3.7.12
+        uses: actions/setup-python@v4
+        with:
+            python-version: "3.7.12"
+      - name: Setup virtual env
+        uses: ./.github/actions/make_init
+      - name: Run make develop
+        run: make develop
+      - name: Register Streamlit User & Mapbox Token
+        run: |
+          if [ ! -d $HOME/.streamlit ] ; then
+            mkdir $HOME/.streamlit
+          fi
+          MAPBOX_TOKEN=$(curl -sS https://data.streamlit.io/tokens.json | jq -r '.["mapbox-localhost"]') \
+          && echo '[mapbox]' >  ~/.streamlit/config.toml \
+          && echo 'token = "'$MAPBOX_TOKEN'"' >> ~/.streamlit/config.toml
+          
+          echo '[general]' >  ~/.streamlit/credentials.toml
+          echo 'email = "test@streamlit.io"' >> ~/.streamlit/credentials.toml 
+      - name: Install Cypress dependencies
+        run: |
+          sudo apt-get install -y xvfb libgtk2.0-0 libnotify-dev libgconf-2-4 libnss3 libxss1 libasound2 jq curl
+      - name: Cypress Tests
+        run: |
+          js_specs=(e2e/specs/*)
+          # get how many specs to be split up into 21 runs
+          (( interval=(${#js_specs[@]}+21-1)/21 ))
+          (( startIndex = (${{matrix.specs}} - 1) * ${interval} ))
```

Comment:
This looks roughly correct to me, but I'll have to go back and double-check it because I don't know file globbing / bash arrays nearly well enough to know at a glance

Reply:
I wonder if we should write this block of code in Python. This would improve its readability and its confidence.